scipy/_lib/_array_api.py

```diff
@@ -191,33 +192,81 @@ def is_torch(xp):
     return xp.__name__ == 'scipy._lib.array_api_compat.array_api_compat.torch'
 
 
-def assert_equal(actual, desired, err_msg='', xp=None):
+def _strict_check(actual, desired, check_xp=True, check_dtype=True, check_shape=True):
+    try:
+        actual.dtype
+        desired.dtype
+    except AttributeError:
+        raise ValueError("Inputs should be arrays with a dtype attribute, "
+                         "python scalars and arrays are not accepted.")
+    if check_xp:
+        _assert_matching_namespace(actual, desired)
+    if check_dtype:
+        assert_(actual.dtype == desired.dtype,
+                f"Desired dtype: {desired.dtype}, actual dtype: {actual.dtype}")
+    if check_shape:
+        msg = (f"Array shapes are not equal: actual shape = {actual.shape}, "
+               f"desired shape = {desired.shape}")
+        assert_(actual.shape == desired.shape, msg)
+
+
+def _assert_matching_namespace(actual, desired):
+    desired_space = array_namespace(desired)
+    if isinstance(actual, tuple):
+        for arr in actual:
+            arr_space = array_namespace(arr)
+            assert arr_space == desired_space
+            if is_numpy(arr_space):
+                _check_scalar(actual, desired)
+    else:
+        actual_space = array_namespace(actual)
+        assert_(actual_space == desired_space,
```

Comment:
I've heard preferences for plain `assert` these days. I'm not sure if that is still true when you want the assertion to have a custom message, but you *can* produce a custom message like `assert 0, "message"`.

Reply:
Never mind, I see that you can't do plain `assert` with implicit line break.